File: 02_Sort_Application/Level3/1931.java
// Copyright@2023 Jihoon Lucas Kim <jihoon.lucas.kim@gmail.com>
//회의실 배정
//https://www.acmicpc.net/problem/1931
//힌트
//1. 끝나는 시간 순으로 정렬
//2. 끝나는 시간이 같으면 시작시간이 빠른 순으로 정렬
//3. 끝나는 순서가 빠른 순으로 예약하면서 예약 가능한 회의만 카운트 해줌.

import java.util.Collections;
import java.util.Scanner;
import java.util.ArrayList;

// 시작시간과 종료시간을 가진 class
class Conf implements Comparable<Conf> {
	public Conf(int startTime, int endTime) {
		this.startTime = startTime;
		this.endTime = endTime;
	}
	
	int startTime;
	int endTime;
	
	@Override
	public int compareTo(Conf o) {
		if (this.endTime > o.endTime) {
			return 1;
		} else if (this.endTime == o.endTime) {
			if (this.startTime > o.startTime) {
				return 1;
			}
		}
		
		return -1;
	}
}

public class Main {
	public static void main(String[] args) {
		Scanner sc = new Scanner(System.in);
		
		int N = sc.nextInt();
		ArrayList<Conf> confArr = new ArrayList<>();
		
		for (int i = 0; i < N; i++) {
			int startTime = sc.nextInt();
			int endTime = sc.nextInt();
			
			confArr.add(new Conf(startTime, endTime));
		}
		
		Collections.sort(confArr);
		
		int currentEndTime = 0;
		int answer = 0;
	    // 앞 회의와 겹치지 않는지 체크하면서 정답 수를 추가 해줌.		
		for (int i = 0; i < N; i++) {
			if (currentEndTime <= confArr.get(i).startTime) {
				currentEndTime = confArr.get(i).endTime;
				answer++;
			}
		}
		
		System.out.println(answer);
	}
}

File: 02_Sort_Application/Level3/1931.cpp
// Copyright@2023 Jihoon Lucas Kim <jihoon.lucas.kim@gmail.com>
// 회의실 배정
// https://www.acmicpc.net/problem/1931
// 힌트
// 1. 끝나는 시간 순으로 정렬
// 2. 끝나는 시간이 같으면 시작시간이 빠른 순으로 정렬
// 3. 끝나는 순서가 빠른 순으로 예약하면서 예약 가능한 회의만 카운트 해줌.
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

bool cmp(pair<int, int> a, pair<int, int> b)
{
    if (a.second == b.second)
        return a.first < b.first;
    else
        return a.second < b.second;
}

int main()
{
    int n;
    cin >> n;
    vector<pair<int, int> > times(n);

    for (int i = 0; i < n; i++)
    {
        cin >> times[i].first >> times[i].second;
    }
    sort(times.begin(), times.end(), cmp);
    // 1. 첫 번째 회의를 먼저 등록하면서 끝나는 시간을 갱신함
    int answer = 1;
    int end_time = times[0].second;
    // 2. 두 번재 회의부터 앞 회의와 겹치지 않는지 체크하면서 정답 수를 추가 해줌.
    for (int i = 1; i < n; i++)
    {
        if (end_time <= times[i].first)
        {
            end_time = times[i].second;
            answer += 1;
        }
    }
    cout << answer << endl;
    return 0;
}